File: backend/app/services/audio_service.py
"""
Audio service - Xử lý phát nhạc
"""
from pathlib import Path
from typing import Optional
import backend.app.config as config


class AudioService:
    """Service cho audio"""
    
    def __init__(self):
        self.audio_dir = config.DATA_DIR / "audio"
    
    def get_command_audio(self, command: str) -> Optional[Path]:
        """
        Lấy đường dẫn audio cho lệnh
        
        Args:
            command: Tên lệnh (vd: "nghiem_di_deu_buoc")
            
        Returns:
            Đường dẫn file audio hoặc None
        """
        audio_path = self.audio_dir / "command" / f"{command}.mp3"
        if audio_path.exists():
            return audio_path
        return None
    
    def get_mode_audio(
        self, 
        criteria: str,  # di_deu hoặc di_nghiem
        mode: str,  # local hoặc global
        practice_type: str  # testing hoặc practising
    ) -> Optional[Path]:
        """
        Lấy đường dẫn audio cho chế độ
        
        Args:
            criteria: Tiêu chí (di_deu/di_nghiem)
            mode: Chế độ (local/global)
            practice_type: Loại (testing/practising)
            
        Returns:
            Đường dẫn file audio hoặc None
        """
        if mode == "local":
            suffix = "short" if practice_type == "testing" else "long"
            audio_path = self.audio_dir / criteria / "local" / f"practising_{suffix}.mp3"
        else:  # global
            audio_path = self.audio_dir / criteria / "global" / "total.mp3"
        
        if audio_path.exists():
            return audio_path
        return None

File: frontend/src/services/audioService.ts
/**
 * Audio service - Xử lý phát nhạc
 */
import apiClient from './api';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

// Lấy URL audio cho lệnh
export const getCommandAudio = (command: string): string | null => {
  return `${API_BASE_URL}/static/audio/command/${command}.mp3`;
};

// Lấy URL audio cho mode
export const getModeAudio = (
  criteria: string,  // di_deu hoặc di_nghiem
  mode: string,  // local hoặc global
  practiceType: string  // testing hoặc practising
): string | null => {
  if (mode === 'local') {
    const suffix = practiceType === 'testing' ? 'short' : 'long';
    return `${API_BASE_URL}/static/audio/${criteria}/local/practising_${suffix}.mp3`;
  } else {
    return `${API_BASE_URL}/static/audio/${criteria}/global/total.mp3`;
  }
};

// Phát audio
export const playAudio = (audioUrl: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    const audio = new Audio(audioUrl);
    audio.onended = () => resolve();
    audio.onerror = () => reject(new Error('Lỗi phát audio'));
    audio.play();
  });
};

export const audioService = { getCommandAudio, getModeAudio, playAudio };
export default audioService;

